### DISCUSSION TITLE
Revamp "sample" function to be a __call__ method

### DISCUSSION THREAD

**Discussion by ConnorStoneAstro** [Autostronomy/AstroPhot#32] [Ideas]
This would be a major refactor of the interface. The idea is that instead of calling:
`model.sample(image)`
one would simply call:
`model(image)`
This is somewhat cleaner, but also makes it clearer what the core purpose of the model is for, sampling images. Internally there could still be a sample method but it would be more low level/functional. For those trying to use AutoProf for a more specialized workflow it may be helpful to have a clear way to call lower level functions as well.  The inspiration comes from pytorch nn.Module which has users implement a `forward` function however the modules themselves are called by `__call__` so that the nn.Module can handle hooks and other housekeeping before `forward` is called. This seems like a useful separation of tasks.

At this point it's just an idea, something to workshop.

**Comment by ranbir7** [comment]
This seems great👍
I'll try looking towards this.

**Reply by ConnorStoneAstro** [reply]
Glad you like the idea! But this is a major refactor that I would want to handle myself. I'd appreciate any ideas you have on how to make this work though! For example, what sorts of housekeeping should be in the __call__ method vs the sample method? When a user writes their own custom models, what should they overload, the sample method, or some other sub function? I was hoping to discuss the idea here before going and making a bunch of changes.

**Reply by ranbir7** [reply]
can u tell me where is the scripts where u will be making changes?
I want to learn it and then I can be suggesting what to do
cuz im still stuck in learning the jacobian issue

**Reply by ConnorStoneAstro** [reply]
It would be a change to the core `AutoProf_Model` object. Subsequently every model, test, and tutorial would need to be adjusted. So it is a major refactor.

**Comment by ConnorStoneAstro** [comment]
Added a new branch `dev-newsample` which shows what the new interface would be like. It's not such a big change as I originally thought, but it does seem pretty nice.

**Comment by ConnorStoneAstro** [comment]
Ok, the new interface has been merged, all the tests have passed, looks good!
